lib/std/tasks.nim

```diff
@@ -0,0 +1,480 @@
+          addAllNode()
+        elif param[0].eqIdent("typeDesc"):
+          callNode.add nnkExprEqExpr.newTree(formalParams[i][0], e[i])
+        elif param[0].eqIdent("varargs") or param[0].eqIdent("openArray"):
+          let
+            seqType = nnkBracketExpr.newTree(newIdentNode("seq"), param[1])
+            scratchDotExpr = newDotExpr(scratchIdent, formalParams[i][0])
+            seqCallNode = newcall("@", e[i])
+
+          case e[i].kind
+          of nnkSym:
+            scratchAssignList.add newCall(newIdentNode("=sink"), scratchDotExpr, seqCallNode)
+          else:
+            scratchAssignList.add newAssignment(scratchDotExpr, seqCallNode)
+
+          let tempNode = genSym(kind = nskTemp)
+          callNode.add nnkExprEqExpr.newTree(formalParams[i][0], tempNode)
+          tempAssignList.add newLetStmt(tempNode, newCall(transferProc, newDotExpr(objTemp, formalParams[i][0])))
+          scratchRecList.add newIdentDefs(newIdentNode(formalParams[i][0].strVal), seqType)
+        else:
+          scratchRecList.add newIdentDefs(newIdentNode(formalParams[i][0].strVal), param)
+          addAllNode()
+      of nnkBracket, nnkObjConstr:
+        # passing by static parameters
+        # so we pass them directly instead of passing by scratchObj
+        callNode.add nnkExprEqExpr.newTree(formalParams[i][0], e[i])
+      of nnkSym, nnkPtrTy:
+        scratchRecList.add newIdentDefs(newIdentNode(formalParams[i][0].strVal), param)
+        addAllNode()
+      of nnkCharLit..nnkNilLit:
+        # TODO params doesn't work for static string
+        callNode.add nnkExprEqExpr.newTree(formalParams[i][0], e[i])
+      else:
+        error("not supported type kinds")
+
+    let stmtList = newStmtList()
+    let scratchObjType = genSym(kind = nskType, ident = "ScratchObj")
+    let scratchObj = nnkTypeSection.newTree(
+                      nnkTypeDef.newTree(
+                        scratchObjType,
+                        newEmptyNode(),
+                        nnkObjectTy.newTree(
+                          newEmptyNode(),
+                          newEmptyNode(),
+                          scratchRecList
+                        )
+                      )
+                    )
+
+
+    let scratchVarSection = nnkVarSection.newTree(
+      nnkIdentDefs.newTree(
+        scratchIdent,
+        scratchObjType,
+        newEmptyNode()
+      )
+    )
+
+    stmtList.add(scratchObj)
+    stmtList.add(scratchVarSection)
+    stmtList.add(scratchAssignList)
+
+    var functionStmtList = newStmtList()
+    let funcCall = newCall(e[0], callNode)
+    functionStmtList.add tempAssignList
+    functionStmtList.add funcCall
+
+    let funcName = genSym(nskProc, e[0].strVal)
+
+    result = quote do:
+      `stmtList`
+
+      proc `funcName`(args: pointer) {.nimcall.} =
+        let `objTemp` = cast[ptr `scratchObjType`](args)
+        `functionStmtList`
+
+      Task(callback: `funcName`, args: addr(`scratchIdent`))
```

Comment:
This cannot work reliably, the storage for task arguments must be put on the heap.

Reply:
what's the type of `args` should be? How can I put them on heap?

Reply:
fixed